Some Formula tests are flaky

### What happened?

```
    async def test_3_phase_formulas(self, mocker: MockerFixture) -> None:
        """Test 3 phase formulas current formulas and their composition."""
        mockgrid = MockMicrogrid(grid_side_meter=False)
        mockgrid.add_batteries(3)
        mockgrid.add_ev_chargers(1)
        request_chan, channel_registry = await mockgrid.start(mocker)
        logical_meter = LogicalMeter(
            channel_registry,
            request_chan.new_sender(),
            microgrid.get().component_graph,
        )
        grid_current_recv = await logical_meter.grid_current()
        pool = EVChargerPool(channel_registry, request_chan.new_sender())
        ev_current_recv = await pool.total_current()
    
        engine = (grid_current_recv.clone() - ev_current_recv.clone()).build(
            "net_current"
        )
        net_current_recv = engine.new_receiver()
    
        await synchronize_receivers(
            [grid_current_recv, ev_current_recv, net_current_recv]
        )
    
        for _ in range(10):
            grid_amps = await grid_current_recv.receive()
            ev_amps = await ev_current_recv.receive()
            net_amps = await net_current_recv.receive()
    
>           assert grid_amps.value_p1 is not None and grid_amps.value_p1 > 0.0
E           assert (None is not None)
E            +  where None = Sample3Phase(timestamp=datetime.datetime(2023, 3, 10, 7, 9, 39, 78987, tzinfo=datetime.timezone.utc), value_p1=None, value_p2=None, value_p3=None).value_p1
```

### What did you expect instead?

We should fix them

### Affected version(s)

_No response_

### Affected part(s)

Unit, integration and performance tests (part:tests)

### Extra information

_No response_

This test is also flaky: 

```
=================================== FAILURES ===================================
_____________________ TestEVChargerPool.test_state_updates _____________________

self = <tests.timeseries.test_ev_charger_pool.TestEVChargerPool object at 0x7ff695e1c990>
mocker = <pytest_mock.plugin.MockerFixture object at 0x7ff68f98a510>

    async def test_state_updates(self, mocker: MockerFixture) -> None:
        """Test ev charger state updates are visible."""
    
        mockgrid = MockMicrogrid(grid_side_meter=False, sample_rate_s=0.01)
        mockgrid.add_ev_chargers(5)
        await mockgrid.start(mocker)
    
        state_tracker = StateTracker(set(mockgrid.evc_ids))
    
        async def check_states(
            expected: dict[int, EVChargerState],
        ) -> None:
            await asyncio.sleep(0.02)
            for comp_id, exp_state in expected.items():
                assert state_tracker.get(comp_id) == exp_state
    
        ## check that all chargers are in idle state.
        expected_states = {evc_id: EVChargerState.IDLE for evc_id in mockgrid.evc_ids}
        assert len(expected_states) == 5
>       await check_states(expected_states)

tests/timeseries/test_ev_charger_pool.py:56: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

expected = {6: <EVChargerState.IDLE: 'IDLE'>, 16: <EVChargerState.IDLE: 'IDLE'>, 26: <EVChargerState.IDLE: 'IDLE'>, 36: <EVChargerState.IDLE: 'IDLE'>, ...}

    async def check_states(
        expected: dict[int, EVChargerState],
    ) -> None:
        await asyncio.sleep(0.02)
        for comp_id, exp_state in expected.items():
>           assert state_tracker.get(comp_id) == exp_state
E           AssertionError: assert <EVChargerState.MISSING: 'MISSING'> == <EVChargerState.IDLE: 'IDLE'>
E            +  where <EVChargerState.MISSING: 'MISSING'> = <bound method StateTracker.get of <frequenz.sdk.timeseries.ev_charger_pool._state_tracker.StateTracker object at 0x7ff68fcc1cd0>>(6)
E            +    where <bound method StateTracker.get of <frequenz.sdk.timeseries.ev_charger_pool._state_tracker.StateTracker object at 0x7ff68fcc1cd0>> = <frequenz.sdk.timeseries.ev_charger_pool._state_tracker.StateTracker object at 0x7ff68fcc1cd0>.get

tests/timeseries/test_ev_charger_pool.py:51: AssertionError
```

--- a/src/frequenz/sdk/timeseries/_formula_engine/_formula_engine_pool.py
+++ b/src/frequenz/sdk/timeseries/_formula_engine/_formula_engine_pool.py
@@ -110,3 +110,8 @@ class FormulaEnginePool:
         ).generate()
         self._engines[channel_key] = engine
         return engine
+
+    async def stop(self) -> None:
+        """Stop all formula engines in the pool."""
+        for engine in self._engines.values():
+            await engine._stop()  # pylint: disable=protected-access

--- a/src/frequenz/sdk/timeseries/_formula_engine/_formula_generators/_pv_power_formula.py
+++ b/src/frequenz/sdk/timeseries/_formula_engine/_formula_generators/_pv_power_formula.py
@@ -3,12 +3,20 @@
 
 """Formula generator for PV Power, from the component graph."""
 
+from __future__ import annotations
+
 import logging
+from collections import abc
 
 from ....microgrid import connection_manager
 from ....microgrid.component import ComponentCategory, ComponentMetricId, InverterType
 from .._formula_engine import FormulaEngine
-from ._formula_generator import NON_EXISTING_COMPONENT_ID, FormulaGenerator, FormulaType
+from ._formula_generator import (
+    NON_EXISTING_COMPONENT_ID,
+    FormulaGenerationError,
+    FormulaGenerator,
+    FormulaType,
+)
 
 _logger = logging.getLogger(__name__)
 
@@ -27,15 +35,8 @@ class PVPowerFormula(FormulaGenerator):
         """
         builder = self._get_builder("pv-power", ComponentMetricId.ACTIVE_POWER)
 
-        component_graph = connection_manager.get().component_graph
-        pv_inverters = list(
-            comp
-            for comp in component_graph.components()
-            if comp.category == ComponentCategory.INVERTER
-            and comp.type == InverterType.SOLAR
-        )
-
-        if not pv_inverters:
+        pv_meters = self._get_pv_meters()
+        if not pv_meters:
             _logger.warning(
                 "Unable to find any PV inverters in the component graph. "
                 "Subscribing to the resampling actor with a non-existing "
@@ -51,11 +52,11 @@ class PVPowerFormula(FormulaGenerator):
 
         builder.push_oper("(")
         builder.push_oper("(")
-        for idx, comp in enumerate(pv_inverters):
+        for idx, comp_id in enumerate(pv_meters):
             if idx > 0:
                 builder.push_oper("+")
 
-            builder.push_component_metric(comp.component_id, nones_are_zeros=True)
+            builder.push_component_metric(comp_id, nones_are_zeros=True)
         builder.push_oper(")")
         if self._config.formula_type == FormulaType.PRODUCTION:
             builder.push_oper("*")
@@ -66,3 +67,40 @@ class PVPowerFormula(FormulaGenerator):
             builder.push_clipper(0.0, None)
 
         return builder.build()
+
+    def _get_pv_meters(self) -> abc.Set[int]:
+        component_graph = connection_manager.get().component_graph
+
+        pv_inverters = list(
+            comp
+            for comp in component_graph.components()
+            if comp.category == ComponentCategory.INVERTER
+            and comp.type == InverterType.SOLAR
+        )
+        pv_meters: set[int] = set()
+
+        if not pv_inverters:
+            return pv_meters
+
+        for pv_inverter in pv_inverters:
+            predecessors = component_graph.predecessors(pv_inverter.component_id)
+            if len(predecessors) != 1:
+                raise FormulaGenerationError(
+                    "Expected exactly one predecessor for PV inverter "
+                    f"{pv_inverter.component_id}, but found {len(predecessors)}."
+                )
+            meter = next(iter(predecessors))
+            if meter.category != ComponentCategory.METER:
+                raise FormulaGenerationError(
+                    f"Expected predecessor of PV inverter {pv_inverter.component_id} "
+                    f"to be a meter, but found {meter.category}."
+                )
+            meter_successors = component_graph.successors(meter.component_id)
+            if len(meter_successors) != 1:
+                raise FormulaGenerationError(
+                    f"Expected exactly one successor for meter {meter.component_id}"
+                    f", connected to PV inverter {pv_inverter.component_id}"
+                    f", but found {len(meter_successors)}."
+                )
+            pv_meters.add(meter.component_id)
+        return pv_meters

--- a/src/frequenz/sdk/timeseries/battery_pool/battery_pool.py
+++ b/src/frequenz/sdk/timeseries/battery_pool/battery_pool.py
@@ -9,7 +9,7 @@ import asyncio
 import uuid
 from collections.abc import Set
 from datetime import timedelta
-from typing import Any
+from typing import Any, Awaitable
 
 from frequenz.channels import Receiver, Sender
 
@@ -78,6 +78,7 @@ class BatteryPool:
 
         self._working_batteries: set[int] = set()
 
+        self._update_battery_status_task: asyncio.Task[None] | None = None
         if self._batteries:
             self._update_battery_status_task = asyncio.create_task(
                 self._update_battery_status(batteries_status_receiver)
@@ -266,10 +267,13 @@ class BatteryPool:
 
     async def stop(self) -> None:
         """Stop all pending async tasks."""
-        await asyncio.gather(
-            *[method.stop() for method in self._active_methods.values()],
-            cancel_and_await(self._update_battery_status_task),
-        )
+        tasks_to_stop: list[Awaitable[Any]] = [
+            method.stop() for method in self._active_methods.values()
+        ]
+        tasks_to_stop.append(self._formula_pool.stop())
+        if self._update_battery_status_task:
+            tasks_to_stop.append(cancel_and_await(self._update_battery_status_task))
+        await asyncio.gather(*tasks_to_stop)
 
     def _get_all_batteries(self) -> Set[int]:
         """Get all batteries from the microgrid.

--- a/src/frequenz/sdk/timeseries/ev_charger_pool/_ev_charger_pool.py
+++ b/src/frequenz/sdk/timeseries/ev_charger_pool/_ev_charger_pool.py
@@ -15,6 +15,7 @@ from datetime import timedelta
 
 from frequenz.channels import Broadcast, ChannelClosedError, Receiver, Sender
 
+from ..._internal._asyncio import cancel_and_await
 from ...actor import ChannelRegistry, ComponentMetricRequest
 from ...microgrid import connection_manager
 from ...microgrid.component import ComponentCategory, ComponentMetricId
@@ -271,6 +272,11 @@ class EVChargerPool:
             await self._bounds_setter.stop()
         if self._state_tracker:
             await self._state_tracker.stop()
+        await self._formula_pool.stop()
+        for stream in self._status_streams.values():
+            task, chan = stream
+            await chan.close()
+            await cancel_and_await(task)
 
     async def _get_current_streams(
         self, component_id: int

--- a/src/frequenz/sdk/timeseries/logical_meter/_logical_meter.py
+++ b/src/frequenz/sdk/timeseries/logical_meter/_logical_meter.py
@@ -402,3 +402,7 @@ class LogicalMeter:
         )
         assert isinstance(engine, FormulaEngine)
         return engine
+
+    async def stop(self) -> None:
+        """Stop all formula engines."""
+        await self._formula_pool.stop()

--- a/tests/timeseries/_battery_pool/test_battery_pool.py
+++ b/tests/timeseries/_battery_pool/test_battery_pool.py
@@ -13,7 +13,7 @@ from typing import Any, AsyncIterator, Generic, Iterator, TypeVar
 
 import async_solipsism
 import pytest
-from frequenz.channels import Broadcast, Receiver, Sender
+from frequenz.channels import Receiver, Sender
 from pytest_mock import MockerFixture
 
 from frequenz.sdk import microgrid
@@ -23,8 +23,7 @@ from frequenz.sdk._internal._constants import (
 )
 from frequenz.sdk.actor import ResamplerConfig
 from frequenz.sdk.actor.power_distributing import BatteryStatus
-from frequenz.sdk.microgrid.component import ComponentCategory, ComponentMetricId
-from frequenz.sdk.timeseries import Sample
+from frequenz.sdk.microgrid.component import ComponentCategory
 from frequenz.sdk.timeseries.battery_pool import (
     BatteryPool,
     Bound,
@@ -445,53 +444,24 @@ async def test_battery_pool_power(mocker: MockerFixture) -> None:
     """Test `BatteryPool.{,production,consumption}_power` methods."""
     mockgrid = MockMicrogrid(grid_side_meter=True)
     mockgrid.add_batteries(2)
-    await mockgrid.start(mocker)
-
-    channels: dict[int, Broadcast[Sample]] = {
-        meter_id: Broadcast(f"#{meter_id}")
-        for meter_id in [*mockgrid.meter_ids, *mockgrid.battery_inverter_ids]
-    }
-    senders: list[Sender[Sample]] = [
-        channels[component_id].new_sender()
-        for component_id in mockgrid.battery_inverter_ids
-    ]
-
-    async def send_resampled_data(
-        now: datetime,
-        meter_data: list[float | None],
-    ) -> None:
-        """Send resampled data to the channels."""
-        for sender, value in zip(senders, meter_data):
-            await sender.send(Sample(now, value))
-
-    def mock_resampled_receiver(
-        _1: Any, component_id: int, _2: ComponentMetricId
-    ) -> Receiver[Sample]:
-        return channels[component_id].new_receiver()
-
-    mocker.patch(
-        "frequenz.sdk.timeseries._formula_engine._resampled_formula_builder"
-        ".ResampledFormulaBuilder._get_resampled_receiver",
-        mock_resampled_receiver,
-    )
+    await mockgrid.start_mock_datapipeline(mocker)
 
     battery_pool = microgrid.battery_pool()
     power_receiver = battery_pool.power.new_receiver()
     consumption_receiver = battery_pool.consumption_power.new_receiver()
     production_receiver = battery_pool.production_power.new_receiver()
 
-    now = datetime.now(tz=timezone.utc)
-    await send_resampled_data(now, [2.0, 3.0])
+    await mockgrid.mock_data.send_bat_inverter_power([2.0, 3.0])
     assert (await power_receiver.receive()).value == 5.0
     assert (await consumption_receiver.receive()).value == 5.0
     assert (await production_receiver.receive()).value == 0.0
 
-    await send_resampled_data(now + timedelta(seconds=1), [-2.0, -5.0])
+    await mockgrid.mock_data.send_bat_inverter_power([-2.0, -5.0])
     assert (await power_receiver.receive()).value == -7.0
     assert (await consumption_receiver.receive()).value == 0.0
     assert (await production_receiver.receive()).value == 7.0
 
-    await send_resampled_data(now + timedelta(seconds=3), [2.0, -5.0])
+    await mockgrid.mock_data.send_bat_inverter_power([2.0, -5.0])
     assert (await power_receiver.receive()).value == -3.0
     assert (await consumption_receiver.receive()).value == 0.0
     assert (await production_receiver.receive()).value == 3.0

--- a/tests/timeseries/_formula_engine/test_formula_composition.py
+++ b/tests/timeseries/_formula_engine/test_formula_composition.py
@@ -4,7 +4,7 @@
 """Tests for formula composition."""
 
 
-from math import isclose
+import math
 
 from pytest_mock import MockerFixture
 
@@ -12,7 +12,7 @@ from frequenz.sdk import microgrid
 from frequenz.sdk.microgrid.component import ComponentMetricId
 
 from ..mock_microgrid import MockMicrogrid
-from .utils import get_resampled_stream, synchronize_receivers
+from .utils import get_resampled_stream
 
 
 class TestFormulaComposition:
@@ -23,10 +23,11 @@ class TestFormulaComposition:
         mocker: MockerFixture,
     ) -> None:
         """Test the composition of formulas."""
-        mockgrid = MockMicrogrid(grid_side_meter=False, sample_rate_s=0.2)
+        mockgrid = MockMicrogrid(grid_side_meter=False)
         mockgrid.add_batteries(3)
         mockgrid.add_solar_inverters(2)
-        await mockgrid.start(mocker)
+        await mockgrid.start_mock_datapipeline(mocker)
+
         logical_meter = microgrid.logical_meter()
         battery_pool = microgrid.battery_pool()
         main_meter_recv = get_resampled_stream(
@@ -41,33 +42,56 @@ class TestFormulaComposition:
         engine = (logical_meter.pv_power + battery_pool.power).build("inv_power")
         inv_calc_recv = engine.new_receiver()
 
-        count = 0
-        for _ in range(10):
-            grid_pow = await grid_power_recv.receive()
-            pv_pow = await pv_power_recv.receive()
-            bat_pow = await battery_power_recv.receive()
-            main_pow = await main_meter_recv.receive()
-            inv_calc_pow = await inv_calc_recv.receive()
+        await mockgrid.mock_data.send_bat_inverter_power([10.0, 12.0, 14.0])
+        await mockgrid.mock_data.send_meter_power(
+            [100.0, 10.0, 12.0, 14.0, -100.0, -200.0]
+        )
 
-            assert grid_pow is not None and grid_pow.value is not None
-            assert inv_calc_pow is not None and inv_calc_pow.value is not None
-            assert bat_pow is not None and bat_pow.value is not None
-            assert pv_pow is not None and pv_pow.value is not None
-            assert main_pow is not None and main_pow.value is not None
-            assert isclose(inv_calc_pow.value, pv_pow.value + bat_pow.value)
-            assert isclose(grid_pow.value, inv_calc_pow.value + main_pow.value)
-            count += 1
+        grid_pow = await grid_power_recv.receive()
+        pv_pow = await pv_power_recv.receive()
+        bat_pow = await battery_power_recv.receive()
+        main_pow = await main_meter_recv.receive()
+        inv_calc_pow = await inv_calc_recv.receive()
+
+        assert (
+            grid_pow is not None
+            and grid_pow.value is not None
+            and math.isclose(grid_pow.value, -164.0)
+        )  # 100 + 10 + 12 + 14 + -100 + -200
+        assert (
+            bat_pow is not None
+            and bat_pow.value is not None
+            and math.isclose(bat_pow.value, 36.0)
+        )  # 10 + 12 + 14
+        assert (
+            pv_pow is not None
+            and pv_pow.value is not None
+            and math.isclose(pv_pow.value, -300.0)
+        )  # -100 + -200
+        assert (
+            inv_calc_pow is not None
+            and inv_calc_pow.value is not None
+            and math.isclose(inv_calc_pow.value, -264.0)  # -300 + 36
+        )
+        assert (
+            main_pow is not None
+            and main_pow.value is not None
+            and math.isclose(main_pow.value, 100.0)
+        )
+
+        assert math.isclose(inv_calc_pow.value, pv_pow.value + bat_pow.value)
+        assert math.isclose(grid_pow.value, inv_calc_pow.value + main_pow.value)
 
         await mockgrid.cleanup()
         await engine._stop()  # pylint: disable=protected-access
-
-        assert count == 10
+        await battery_pool.stop()
+        await logical_meter.stop()
 
     async def test_formula_composition_missing_pv(self, mocker: MockerFixture) -> None:
         """Test the composition of formulas with missing PV power data."""
         mockgrid = MockMicrogrid(grid_side_meter=False)
         mockgrid.add_batteries(3)
-        await mockgrid.start(mocker)
+        await mockgrid.start_mock_datapipeline(mocker)
         battery_pool = microgrid.battery_pool()
         logical_meter = microgrid.logical_meter()
 
@@ -78,6 +102,11 @@ class TestFormulaComposition:
 
         count = 0
         for _ in range(10):
+            await mockgrid.mock_data.send_bat_inverter_power(
+                [10.0 + count, 12.0 + count, 14.0 + count]
+            )
+            await mockgrid.mock_data.send_non_existing_component_value()
+
             bat_pow = await battery_power_recv.receive()
             pv_pow = await pv_power_recv.receive()
             inv_pow = await inv_calc_recv.receive()
@@ -88,6 +117,8 @@ class TestFormulaComposition:
 
         await mockgrid.cleanup()
         await engine._stop()  # pylint: disable=protected-access
+        await battery_pool.stop()
+        await logical_meter.stop()
 
         assert count == 10
 
@@ -95,7 +126,7 @@ class TestFormulaComposition:
         """Test the composition of formulas with missing battery power data."""
         mockgrid = MockMicrogrid(grid_side_meter=False)
         mockgrid.add_solar_inverters(2)
-        await mockgrid.start(mocker)
+        await mockgrid.start_mock_datapipeline(mocker)
         battery_pool = microgrid.battery_pool()
         logical_meter = microgrid.logical_meter()
 
@@ -106,6 +137,10 @@ class TestFormulaComposition:
 
         count = 0
         for _ in range(10):
+            await mockgrid.mock_data.send_meter_power(
+                [10.0 + count, 12.0 + count, 14.0 + count]
+            )
+            await mockgrid.mock_data.send_non_existing_component_value()
             bat_pow = await battery_power_recv.receive()
             pv_pow = await pv_power_recv.receive()
             inv_pow = await inv_calc_recv.receive()
@@ -116,15 +151,19 @@ class TestFormulaComposition:
 
         await mockgrid.cleanup()
         await engine._stop()  # pylint: disable=protected-access
+        await battery_pool.stop()
+        await logical_meter.stop()
 
         assert count == 10
 
     async def test_3_phase_formulas(self, mocker: MockerFixture) -> None:
         """Test 3 phase formulas current formulas and their composition."""
-        mockgrid = MockMicrogrid(grid_side_meter=False, sample_rate_s=0.05)
+        mockgrid = MockMicrogrid(
+            grid_side_meter=False, sample_rate_s=0.05, num_namespaces=2
+        )
         mockgrid.add_batteries(3)
         mockgrid.add_ev_chargers(1)
-        await mockgrid.start(mocker)
+        await mockgrid.start_mock_datapipeline(mocker)
         logical_meter = microgrid.logical_meter()
         ev_pool = microgrid.ev_charger_pool()
 
@@ -134,11 +173,21 @@ class TestFormulaComposition:
         engine = (logical_meter.grid_current - ev_pool.current).build("net_current")
         net_current_recv = engine.new_receiver()
 
-        await synchronize_receivers(
-            [grid_current_recv, ev_current_recv, net_current_recv]
-        )
+        count = 0
 
         for _ in range(10):
+            await mockgrid.mock_data.send_meter_current(
+                [
+                    [10.0, 12.0, 14.0],
+                    [10.0, 12.0, 14.0],
+                    [10.0, 12.0, 14.0],
+                    [10.0, 12.0, 14.0],
+                ]
+            )
+            await mockgrid.mock_data.send_evc_current(
+                [[10.0 + count, 12.0 + count, 14.0 + count]]
+            )
+
             grid_amps = await grid_current_recv.receive()
             ev_amps = await ev_current_recv.receive()
             net_amps = await net_current_recv.receive()
@@ -156,6 +205,11 @@ class TestFormulaComposition:
             assert net_amps.value_p1 == grid_amps.value_p1 - ev_amps.value_p1
             assert net_amps.value_p2 == grid_amps.value_p2 - ev_amps.value_p2
             assert net_amps.value_p3 == grid_amps.value_p3 - ev_amps.value_p3
+            count += 1
 
         await mockgrid.cleanup()
         await engine._stop()  # pylint: disable=protected-access
+        await logical_meter.stop()
+        await ev_pool.stop()
+
+        assert count == 10

--- a/tests/timeseries/_formula_engine/utils.py
+++ b/tests/timeseries/_formula_engine/utils.py
@@ -5,15 +5,13 @@
 
 from __future__ import annotations
 
-from datetime import datetime
 from math import isclose
-from typing import overload
 
 from frequenz.channels import Receiver
 
 from frequenz.sdk.microgrid import _data_pipeline
 from frequenz.sdk.microgrid.component import ComponentMetricId
-from frequenz.sdk.timeseries import Sample, Sample3Phase
+from frequenz.sdk.timeseries import Sample
 from frequenz.sdk.timeseries._formula_engine import ResampledFormulaBuilder
 
 
@@ -48,47 +46,3 @@ def equal_float_lists(list1: list[float], list2: list[float]) -> bool:
         and len(list1) == len(list2)
         and all(isclose(v1, v2) for v1, v2 in zip(list1, list2))
     )
-
-
-@overload
-async def synchronize_receivers(
-    receivers: list[Receiver[Sample]],
-) -> None:
-    ...
-
-
-@overload
-async def synchronize_receivers(
-    receivers: list[Receiver[Sample3Phase]],
-) -> None:
-    ...
-
-
-async def synchronize_receivers(
-    receivers: list[Receiver[Sample]] | list[Receiver[Sample3Phase]],
-) -> None:
-    """Check if given receivers are all returning the same timestamp.
-
-    If not, try to synchronize them.
-    """
-    by_ts: dict[datetime, list[Receiver[Sample] | Receiver[Sample3Phase]]] = {}
-    for recv in receivers:
-        while True:
-            sample = await recv.receive()
-            assert sample is not None
-            if isinstance(sample, Sample) and sample.value is None:
-                continue
-            if isinstance(sample, Sample3Phase) and sample.value_p1 is None:
-                continue
-            by_ts.setdefault(sample.timestamp, []).append(recv)
-            break
-    latest_ts = max(by_ts)
-
-    for sample_ts, recvs in by_ts.items():
-        if sample_ts == latest_ts:
-            continue
-        while sample_ts < latest_ts:
-            for recv in recvs:
-                val = await recv.receive()
-                assert val is not None
-                sample_ts = val.timestamp

--- a/tests/timeseries/mock_datapipeline.py
+++ b/tests/timeseries/mock_datapipeline.py
@@ -1,0 +1,191 @@
+# License: MIT
+# Copyright © 2023 Frequenz Energy-as-a-Service GmbH
+
+"""Mock data_pipeline."""
+
+from __future__ import annotations
+
+import asyncio
+from datetime import datetime
+
+from frequenz.channels import Broadcast, Receiver, Sender
+from pytest_mock import MockerFixture
+
+from frequenz.sdk.actor import ComponentMetricRequest, ResamplerConfig
+from frequenz.sdk.microgrid._data_pipeline import _DataPipeline
+from frequenz.sdk.microgrid.component import ComponentMetricId
+from frequenz.sdk.timeseries import Sample
+from frequenz.sdk.timeseries._formula_engine._formula_generators._formula_generator import (
+    NON_EXISTING_COMPONENT_ID,
+)
+
+# pylint: disable=too-many-instance-attributes
+
+
+class MockDataPipeline:
+    """Mock data_pipeline."""
+
+    def __init__(  # pylint: disable=too-many-arguments
+        self,
+        mocker: MockerFixture,
+        resampler_config: ResamplerConfig,
+        bat_inverter_ids: list[int],
+        pv_inverter_ids: list[int],
+        evc_ids: list[int],
+        chp_ids: list[int],
+        meter_ids: list[int],
+        namespaces: int,
+    ) -> None:
+        """Create a `MockDataPipeline` instance."""
+        self._data_pipeline = _DataPipeline(resampler_config)
+
+        self._channel_registry = self._data_pipeline._channel_registry
+        self._resampler_request_channel = Broadcast[ComponentMetricRequest](
+            "resampler-request"
+        )
+        self._basic_receivers: dict[str, list[Receiver[Sample]]] = {}
+
+        def power_senders(
+            comp_ids: list[int],
+        ) -> list[Sender[Sample]]:
+            senders: list[Sender[Sample]] = []
+            for comp_id in comp_ids:
+                name = f"{comp_id}:{ComponentMetricId.ACTIVE_POWER}"
+                senders.append(self._channel_registry.new_sender(name))
+                self._basic_receivers[name] = [
+                    self._channel_registry.new_receiver(name) for _ in range(namespaces)
+                ]
+            return senders
+
+        self._bat_inverter_power_senders = power_senders(bat_inverter_ids)
+        self._pv_inverter_power_senders = power_senders(pv_inverter_ids)
+        self._ev_power_senders = power_senders(evc_ids)
+        self._chp_power_senders = power_senders(chp_ids)
+        self._meter_power_senders = power_senders(meter_ids)
+        self._non_existing_component_sender = power_senders(
+            [NON_EXISTING_COMPONENT_ID]
+        )[0]
+
+        def current_senders(ids: list[int]) -> list[list[Sender[Sample]]]:
+            senders: list[list[Sender[Sample]]] = []
+            for comp_id in ids:
+                p1_name = f"{comp_id}:{ComponentMetricId.CURRENT_PHASE_1}"
+                p2_name = f"{comp_id}:{ComponentMetricId.CURRENT_PHASE_2}"
+                p3_name = f"{comp_id}:{ComponentMetricId.CURRENT_PHASE_3}"
+
+                senders.append(
+                    [
+                        self._channel_registry.new_sender(p1_name),
+                        self._channel_registry.new_sender(p2_name),
+                        self._channel_registry.new_sender(p3_name),
+                    ]
+                )
+                self._basic_receivers[p1_name] = [
+                    self._channel_registry.new_receiver(p1_name)
+                    for _ in range(namespaces)
+                ]
+                self._basic_receivers[p2_name] = [
+                    self._channel_registry.new_receiver(p2_name)
+                    for _ in range(namespaces)
+                ]
+                self._basic_receivers[p3_name] = [
+                    self._channel_registry.new_receiver(p3_name)
+                    for _ in range(namespaces)
+                ]
+            return senders
+
+        self._bat_inverter_current_senders = current_senders(bat_inverter_ids)
+        self._pv_inverter_current_senders = current_senders(pv_inverter_ids)
+        self._ev_current_senders = current_senders(evc_ids)
+        self._chp_current_senders = current_senders(chp_ids)
+        self._meter_current_senders = current_senders(meter_ids)
+
+        self._next_ts = datetime.now()
+
+        mocker.patch(
+            "frequenz.sdk.microgrid._data_pipeline._DataPipeline"
+            "._resampling_request_sender",
+            self._resampling_request_sender,
+        )
+        mocker.patch(
+            "frequenz.sdk.microgrid._data_pipeline._DATA_PIPELINE", self._data_pipeline
+        )
+
+        self._forward_tasks: dict[str, asyncio.Task[None]] = {}
+        asyncio.create_task(self._handle_resampling_requests())
+
+    def _resampling_request_sender(self) -> Sender[ComponentMetricRequest]:
+        return self._resampler_request_channel.new_sender()
+
+    async def _channel_forward_messages(
+        self, receiver: Receiver[Sample], sender: Sender[Sample]
+    ) -> None:
+        async for sample in receiver:
+            await sender.send(sample)
+
+    async def _handle_resampling_requests(self) -> None:
+        async for request in self._resampler_request_channel.new_receiver():
+            if request.get_channel_name() in self._forward_tasks:
+                continue
+            basic_recv_name = f"{request.component_id}:{request.metric_id}"
+            recv = self._basic_receivers[basic_recv_name].pop()
+            assert recv is not None
+            self._forward_tasks[request.get_channel_name()] = asyncio.create_task(
+                self._channel_forward_messages(
+                    recv,
+                    self._channel_registry.new_sender(request.get_channel_name()),
+                )
+            )
+
+    async def send_meter_power(self, values: list[float | None]) -> None:
+        """Send the given values as resampler output for meter power."""
+        assert len(values) == len(self._meter_power_senders)
+        for chan, value in zip(self._meter_power_senders, values):
+            sample = Sample(self._next_ts, value)
+            await chan.send(sample)
+
+    async def send_evc_power(self, values: list[float | None]) -> None:
+        """Send the given values as resampler output for EV Charger power."""
+        assert len(values) == len(self._ev_power_senders)
+        for chan, value in zip(self._ev_power_senders, values):
+            sample = Sample(self._next_ts, value)
+            await chan.send(sample)
+
+    async def send_bat_inverter_power(self, values: list[float | None]) -> None:
+        """Send the given values as resampler output for battery inverter power."""
+        assert len(values) == len(self._bat_inverter_power_senders)
+        for chan, value in zip(self._bat_inverter_power_senders, values):
+            sample = Sample(self._next_ts, value)
+            await chan.send(sample)
+
+    async def send_non_existing_component_value(self) -> None:
+        """Send a value for a non existing component."""
+        sample = Sample(self._next_ts, None)
+        await self._non_existing_component_sender.send(sample)
+
+    async def send_evc_current(self, values: list[list[float | None]]) -> None:
+        """Send the given values as resampler output for EV Charger current."""
+        assert len(values) == len(self._ev_current_senders)
+        for chan, evc_values in zip(self._ev_current_senders, values):
+            assert len(evc_values) == 3  # 3 values for phases
+            for phase, value in enumerate(evc_values):
+                sample = Sample(self._next_ts, value)
+                await chan[phase].send(sample)
+
+    async def send_bat_inverter_current(self, values: list[list[float | None]]) -> None:
+        """Send the given values as resampler output for battery inverter current."""
+        assert len(values) == len(self._bat_inverter_current_senders)
+        for chan, bat_values in zip(self._bat_inverter_current_senders, values):
+            assert len(bat_values) == 3  # 3 values for phases
+            for phase, value in enumerate(bat_values):
+                sample = Sample(self._next_ts, value)
+                await chan[phase].send(sample)
+
+    async def send_meter_current(self, values: list[list[float | None]]) -> None:
+        """Send the given values as resampler output for meter current."""
+        assert len(values) == len(self._meter_current_senders)
+        for chan, meter_values in zip(self._meter_current_senders, values):
+            assert len(meter_values) == 3  # 3 values for phases
+            for phase, value in enumerate(meter_values):
+                sample = Sample(self._next_ts, value)
+                await chan[phase].send(sample)

--- a/tests/timeseries/mock_microgrid.py
+++ b/tests/timeseries/mock_microgrid.py
@@ -25,14 +25,15 @@ from frequenz.sdk.microgrid.component import (
     EVChargerComponentState,
     InverterType,
 )
-from tests.utils import MockMicrogridClient
 
+from ..utils import MockMicrogridClient
 from ..utils.component_data_wrapper import (
     BatteryDataWrapper,
     EvChargerDataWrapper,
     InverterDataWrapper,
     MeterDataWrapper,
 )
+from .mock_datapipeline import MockDataPipeline
 
 
 class MockMicrogrid:  # pylint: disable=too-many-instance-attributes
@@ -48,16 +49,28 @@ class MockMicrogrid:  # pylint: disable=too-many-instance-attributes
     battery_id_suffix = 9
 
     _microgrid: MockMicrogridClient
+    mock_data: MockDataPipeline
 
-    def __init__(
-        self, grid_side_meter: bool, num_values: int = 2000, sample_rate_s: float = 0.01
+    def __init__(  # pylint: disable=too-many-arguments
+        self,
+        grid_side_meter: bool,
+        api_client_streaming: bool = False,
+        num_values: int = 2000,
+        sample_rate_s: float = 0.01,
+        num_namespaces: int = 1,
     ):
         """Create a new instance.
 
         Args:
             grid_side_meter: whether the main meter should be on the grid side or not.
+            api_client_streaming: whether the mock client should be configured to stream
+                raw data from the API client.
             num_values: number of values to generate for each component.
             sample_rate_s: sample rate in seconds.
+            num_namespaces: number of namespaces that each metric should be available
+                to.  Useful in tests where multiple namespaces (logical_meter,
+                battery_pool, etc) are used, and the same metric is used by formulas in
+                different namespaces.
         """
         self._components: Set[Component] = set(
             [
@@ -68,8 +81,10 @@ class MockMicrogrid:  # pylint: disable=too-many-instance-attributes
         self._connections: Set[Connection] = set([Connection(1, 4)])
         self._id_increment = 0
         self._grid_side_meter = grid_side_meter
+        self._api_client_streaming = api_client_streaming
         self._num_values = num_values
         self._sample_rate_s = sample_rate_s
+        self._namespaces = num_namespaces
 
         self._connect_to = self.grid_id
         if self._grid_side_meter:
@@ -90,10 +105,31 @@ class MockMicrogrid:  # pylint: disable=too-many-instance-attributes
         self._streaming_tasks: list[asyncio.Task[None]] = []
         self._start_meter_streaming(4)
 
+    async def start_mock_datapipeline(self, mocker: MockerFixture) -> None:
+        """Start the MockDataPipeline."""
+        self.init_mock_client(lambda mock_client: mock_client.initialize(mocker))
+        self.mock_data = MockDataPipeline(
+            mocker,
+            ResamplerConfig(timedelta(seconds=self._sample_rate_s)),
+            bat_inverter_ids=self.battery_inverter_ids,
+            pv_inverter_ids=self.pv_inverter_ids,
+            evc_ids=self.evc_ids,
+            meter_ids=self.meter_ids,
+            chp_ids=self.chp_ids,
+            namespaces=self._namespaces,
+        )
+
+    def init_mock_client(
+        self, initialize_cb: Callable[[MockMicrogridClient], None]
+    ) -> None:
+        """Set up the mock client. Does not start the streaming tasks."""
+        self._microgrid = MockMicrogridClient(self._components, self._connections)
+        initialize_cb(self._microgrid)
+
     def start_mock_client(
         self, initialize_cb: Callable[[MockMicrogridClient], None]
     ) -> MockMicrogridClient:
-        """Set up the mock client.
+        """Start the mock client.
 
         Creates the microgrid mock client, initializes it, and starts the streaming
         tasks.
@@ -106,10 +142,7 @@ class MockMicrogrid:  # pylint: disable=too-many-instance-attributes
         Returns:
             A MockMicrogridClient instance.
         """
-        self._microgrid = MockMicrogridClient(self._components, self._connections)
-
-        initialize_cb(self._microgrid)
-
+        self.init_mock_client(initialize_cb)
         self._streaming_tasks = [
             asyncio.create_task(coro) for coro in self._streaming_coros
         ]
@@ -150,6 +183,8 @@ class MockMicrogrid:  # pylint: disable=too-many-instance-attributes
         await self._microgrid.close_channel(comp_id)
 
     def _start_meter_streaming(self, meter_id: int) -> None:
+        if not self._api_client_streaming:
+            return
         self._streaming_coros.append(
             self._comp_data_send_task(
                 meter_id,
@@ -163,6 +198,8 @@ class MockMicrogrid:  # pylint: disable=too-many-instance-attributes
         )
 
     def _start_battery_streaming(self, bat_id: int) -> None:
+        if not self._api_client_streaming:
+            return
         self._streaming_coros.append(
             self._comp_data_send_task(
                 bat_id,
@@ -173,6 +210,8 @@ class MockMicrogrid:  # pylint: disable=too-many-instance-attributes
         )
 
     def _start_inverter_streaming(self, inv_id: int) -> None:
+        if not self._api_client_streaming:
+            return
         self._streaming_coros.append(
             self._comp_data_send_task(
                 inv_id,
@@ -183,6 +222,8 @@ class MockMicrogrid:  # pylint: disable=too-many-instance-attributes
         )
 
     def _start_ev_charger_streaming(self, evc_id: int) -> None:
+        if not self._api_client_streaming:
+            return
         self._streaming_coros.append(
             self._comp_data_send_task(
                 evc_id,

--- a/tests/timeseries/test_ev_charger_pool.py
+++ b/tests/timeseries/test_ev_charger_pool.py
@@ -6,19 +6,14 @@
 from __future__ import annotations
 
 import asyncio
-from datetime import datetime, timezone
-from typing import Any
 
-from frequenz.channels import Broadcast, Receiver, Sender
 from pytest_mock import MockerFixture
 
 from frequenz.sdk import microgrid
 from frequenz.sdk.microgrid.component import (
-    ComponentMetricId,
     EVChargerCableState,
     EVChargerComponentState,
 )
-from frequenz.sdk.timeseries import Sample
 from frequenz.sdk.timeseries.ev_charger_pool._state_tracker import (
     EVChargerState,
     StateTracker,
@@ -32,9 +27,11 @@ class TestEVChargerPool:
     async def test_state_updates(self, mocker: MockerFixture) -> None:
         """Test ev charger state updates are visible."""
 
-        mockgrid = MockMicrogrid(grid_side_meter=False, sample_rate_s=0.01)
+        mockgrid = MockMicrogrid(
+            grid_side_meter=False, api_client_streaming=True, sample_rate_s=0.01
+        )
         mockgrid.add_ev_chargers(5)
-        await mockgrid.start(mocker)
+        mockgrid.start_mock_client(lambda client: client.initialize(mocker))
 
         state_tracker = StateTracker(set(mockgrid.evc_ids))
 
@@ -81,47 +78,19 @@ class TestEVChargerPool:
         """Test the ev power formula."""
         mockgrid = MockMicrogrid(grid_side_meter=False)
         mockgrid.add_ev_chargers(3)
-        await mockgrid.start(mocker)
-
-        channels: dict[int, Broadcast[Sample]] = {
-            meter_id: Broadcast(f"#{meter_id}")
-            for meter_id in [*mockgrid.meter_ids, *mockgrid.evc_ids]
-        }
-        senders: list[Sender[Sample]] = [
-            channels[component_id].new_sender() for component_id in mockgrid.evc_ids
-        ]
-
-        async def send_resampled_data(
-            now: datetime,
-            meter_data: list[float | None],
-        ) -> None:
-            """Send resampled data to the channels."""
-            for sender, value in zip(senders, meter_data):
-                await sender.send(Sample(now, value))
-
-        def mock_resampled_receiver(
-            _1: Any, component_id: int, _2: ComponentMetricId
-        ) -> Receiver[Sample]:
-            return channels[component_id].new_receiver()
-
-        mocker.patch(
-            "frequenz.sdk.timeseries._formula_engine._resampled_formula_builder"
-            ".ResampledFormulaBuilder._get_resampled_receiver",
-            mock_resampled_receiver,
-        )
+        await mockgrid.start_mock_datapipeline(mocker)
 
         ev_pool = microgrid.ev_charger_pool()
         power_receiver = ev_pool.power.new_receiver()
         production_receiver = ev_pool.production_power.new_receiver()
         consumption_receiver = ev_pool.consumption_power.new_receiver()
 
-        now = datetime.now(tz=timezone.utc)
-        await send_resampled_data(now, [2.0, 4.0, 10.0])
+        await mockgrid.mock_data.send_evc_power([2.0, 4.0, 10.0])
         assert (await power_receiver.receive()).value == 16.0
         assert (await production_receiver.receive()).value == 0.0
         assert (await consumption_receiver.receive()).value == 16.0
 
-        await send_resampled_data(now, [2.0, 4.0, -10.0])
+        await mockgrid.mock_data.send_evc_power([2.0, 4.0, -10.0])
         assert (await power_receiver.receive()).value == -4.0
         assert (await production_receiver.receive()).value == 4.0
         assert (await consumption_receiver.receive()).value == 0.0
@@ -130,48 +99,22 @@ class TestEVChargerPool:
 
     async def test_ev_component_data(self, mocker: MockerFixture) -> None:
         """Test the component_data method of EVChargerPool."""
-        mockgrid = MockMicrogrid(grid_side_meter=False)
+        mockgrid = MockMicrogrid(grid_side_meter=False, api_client_streaming=True)
         mockgrid.add_ev_chargers(1)
-        await mockgrid.start(mocker)
+
+        # The component_data method is a bit special because it uses both raw data
+        # coming from the MicrogridClient to get the component state, and resampler data
+        # to get the 3-phase current values.  So we need both `start_mock_client` and
+        # `start_mock_datapipeline`.
+        mockgrid.start_mock_client(lambda client: client.initialize(mocker))
+        await mockgrid.start_mock_datapipeline(mocker)
+
         evc_id = mockgrid.evc_ids[0]
-
         ev_pool = microgrid.ev_charger_pool()
-
-        resampled_p1_channel = Broadcast[Sample]("resampled-current-phase-1")
-        resampled_p2_channel = Broadcast[Sample]("resampled-current-phase-2")
-        resampled_p3_channel = Broadcast[Sample]("resampled-current-phase-3")
-
-        async def send_resampled_current(
-            phase_1: float | None, phase_2: float | None, phase_3: float | None
-        ) -> None:
-            sender_p1 = resampled_p1_channel.new_sender()
-            sender_p2 = resampled_p2_channel.new_sender()
-            sender_p3 = resampled_p3_channel.new_sender()
-
-            now = datetime.now()
-            asyncio.gather(
-                sender_p1.send(Sample(now, phase_1)),
-                sender_p2.send(Sample(now, phase_2)),
-                sender_p3.send(Sample(now, phase_3)),
-            )
-
-        async def mock_current_streams(
-            _1: Any, _2: int
-        ) -> tuple[Receiver[Sample], Receiver[Sample], Receiver[Sample]]:
-            return (
-                resampled_p1_channel.new_receiver(),
-                resampled_p2_channel.new_receiver(),
-                resampled_p3_channel.new_receiver(),
-            )
-
-        mocker.patch(
-            "frequenz.sdk.timeseries.ev_charger_pool.EVChargerPool._get_current_streams",
-            mock_current_streams,
-        )
 
         recv = ev_pool.component_data(evc_id)
 
-        await send_resampled_current(2, 3, 5)
+        await mockgrid.mock_data.send_evc_current([[2, 3, 5]])
         await asyncio.sleep(0.02)
         status = await recv.receive()
         assert (
@@ -181,7 +124,7 @@ class TestEVChargerPool:
         ) == (2, 3, 5)
         assert status.state == EVChargerState.MISSING
 
-        await send_resampled_current(2, 3, None)
+        await mockgrid.mock_data.send_evc_current([[2, 3, None]])
         await asyncio.sleep(0.02)
         status = await recv.receive()
         assert (
@@ -191,7 +134,7 @@ class TestEVChargerPool:
         ) == (2, 3, None)
         assert status.state == EVChargerState.IDLE
 
-        await send_resampled_current(None, None, None)
+        await mockgrid.mock_data.send_evc_current([[None, None, None]])
         await asyncio.sleep(0.02)
         status = await recv.receive()
         assert (
@@ -201,7 +144,7 @@ class TestEVChargerPool:
         ) == (None, None, None)
         assert status.state == EVChargerState.MISSING
 
-        await send_resampled_current(None, None, None)
+        await mockgrid.mock_data.send_evc_current([[None, None, None]])
         mockgrid.evc_cable_states[evc_id] = EVChargerCableState.EV_PLUGGED
         await asyncio.sleep(0.02)
         status = await recv.receive()
@@ -212,7 +155,7 @@ class TestEVChargerPool:
         ) == (None, None, None)
         assert status.state == EVChargerState.MISSING
 
-        await send_resampled_current(4, None, None)
+        await mockgrid.mock_data.send_evc_current([[4, None, None]])
         await asyncio.sleep(0.02)
         status = await recv.receive()
         assert (

--- a/tests/timeseries/test_ev_charger_pool.py
+++ b/tests/timeseries/test_ev_charger_pool.py
@@ -99,7 +99,11 @@ class TestEVChargerPool:
 
     async def test_ev_component_data(self, mocker: MockerFixture) -> None:
         """Test the component_data method of EVChargerPool."""
-        mockgrid = MockMicrogrid(grid_side_meter=False, api_client_streaming=True)
+        mockgrid = MockMicrogrid(
+            grid_side_meter=False,
+            api_client_streaming=True,
+            sample_rate_s=0.05,
+        )
         mockgrid.add_ev_chargers(1)
 
         # The component_data method is a bit special because it uses both raw data
@@ -115,7 +119,7 @@ class TestEVChargerPool:
         recv = ev_pool.component_data(evc_id)
 
         await mockgrid.mock_data.send_evc_current([[2, 3, 5]])
-        await asyncio.sleep(0.02)
+        await asyncio.sleep(0.1)
         status = await recv.receive()
         assert (
             status.current.value_p1,
@@ -125,7 +129,7 @@ class TestEVChargerPool:
         assert status.state == EVChargerState.MISSING
 
         await mockgrid.mock_data.send_evc_current([[2, 3, None]])
-        await asyncio.sleep(0.02)
+        await asyncio.sleep(0.1)
         status = await recv.receive()
         assert (
             status.current.value_p1,
@@ -135,7 +139,7 @@ class TestEVChargerPool:
         assert status.state == EVChargerState.IDLE
 
         await mockgrid.mock_data.send_evc_current([[None, None, None]])
-        await asyncio.sleep(0.02)
+        await asyncio.sleep(0.1)
         status = await recv.receive()
         assert (
             status.current.value_p1,
@@ -146,7 +150,7 @@ class TestEVChargerPool:
 
         await mockgrid.mock_data.send_evc_current([[None, None, None]])
         mockgrid.evc_cable_states[evc_id] = EVChargerCableState.EV_PLUGGED
-        await asyncio.sleep(0.02)
+        await asyncio.sleep(0.1)
         status = await recv.receive()
         assert (
             status.current.value_p1,
@@ -156,7 +160,7 @@ class TestEVChargerPool:
         assert status.state == EVChargerState.MISSING
 
         await mockgrid.mock_data.send_evc_current([[4, None, None]])
-        await asyncio.sleep(0.02)
+        await asyncio.sleep(0.1)
         status = await recv.receive()
         assert (
             status.current.value_p1,
@@ -166,3 +170,4 @@ class TestEVChargerPool:
         assert status.state == EVChargerState.EV_PLUGGED
 
         await mockgrid.cleanup()
+        await ev_pool.stop()

--- a/tests/timeseries/test_logical_meter.py
+++ b/tests/timeseries/test_logical_meter.py
@@ -5,21 +5,12 @@
 
 from __future__ import annotations
 
-from datetime import datetime, timezone
-from typing import Any
-
-from frequenz.channels import Broadcast, Receiver, Sender
 from pytest_mock import MockerFixture
 
 from frequenz.sdk import microgrid
 from frequenz.sdk.microgrid.component import ComponentMetricId
-from frequenz.sdk.timeseries import Sample
 
-from ._formula_engine.utils import (
-    equal_float_lists,
-    get_resampled_stream,
-    synchronize_receivers,
-)
+from ._formula_engine.utils import equal_float_lists, get_resampled_stream
 from .mock_microgrid import MockMicrogrid
 
 # pylint: disable=too-many-locals
@@ -33,7 +24,7 @@ class TestLogicalMeter:
         mockgrid = MockMicrogrid(grid_side_meter=True)
         mockgrid.add_batteries(2)
         mockgrid.add_solar_inverters(1)
-        await mockgrid.start(mocker)
+        await mockgrid.start_mock_datapipeline(mocker)
         logical_meter = microgrid.logical_meter()
 
         grid_power_recv = logical_meter.grid_power.new_receiver()
@@ -44,12 +35,12 @@ class TestLogicalMeter:
             ComponentMetricId.ACTIVE_POWER,
         )
 
-        await synchronize_receivers([grid_power_recv, main_meter_recv])
         results = []
         main_meter_data = []
-        for _ in range(10):
+        for count in range(10):
+            await mockgrid.mock_data.send_meter_power([20.0 + count, 12.0, -13.0, -5.0])
             val = await main_meter_recv.receive()
-            assert val is not None and val.value is not None and val.value > 0.0
+            assert val is not None and val.value is not None and val.value != 0.0
             main_meter_data.append(val.value)
 
             val = await grid_power_recv.receive()
@@ -66,7 +57,7 @@ class TestLogicalMeter:
         mockgrid = MockMicrogrid(grid_side_meter=False)
         mockgrid.add_batteries(2)
         mockgrid.add_solar_inverters(1)
-        await mockgrid.start(mocker)
+        await mockgrid.start_mock_datapipeline(mocker)
         logical_meter = microgrid.logical_meter()
 
         grid_power_recv = logical_meter.grid_power.new_receiver()
@@ -80,15 +71,14 @@ class TestLogicalMeter:
             for meter_id in mockgrid.meter_ids
         ]
 
-        await synchronize_receivers([grid_power_recv, *meter_receivers])
-
         results = []
         meter_sums = []
-        for _ in range(10):
+        for count in range(10):
+            await mockgrid.mock_data.send_meter_power([20.0 + count, 12.0, -13.0, -5.0])
             meter_sum = 0.0
             for recv in meter_receivers:
                 val = await recv.receive()
-                assert val is not None and val.value is not None and val.value > 0.0
+                assert val is not None and val.value is not None and val.value != 0.0
                 meter_sum += val.value
 
             val = await grid_power_recv.receive()
@@ -109,46 +99,19 @@ class TestLogicalMeter:
         mockgrid = MockMicrogrid(grid_side_meter=False)
         mockgrid.add_batteries(2)
         mockgrid.add_solar_inverters(1)
-        await mockgrid.start(mocker)
-
-        channels: dict[int, Broadcast[Sample]] = {
-            meter_id: Broadcast(f"#{meter_id}") for meter_id in mockgrid.meter_ids
-        }
-        senders: list[Sender[Sample]] = [
-            channels[meter_id].new_sender() for meter_id in mockgrid.meter_ids
-        ]
-
-        async def send_resampled_data(
-            now: datetime,
-            meter_data: tuple[float | None, float | None, float | None, float | None],
-        ) -> None:
-            """Send resampled data to the channels."""
-            for sender, value in zip(senders, meter_data):
-                await sender.send(Sample(now, value))
-
-        def mock_resampled_receiver(
-            _1: Any, component_id: int, _2: ComponentMetricId
-        ) -> Receiver[Sample]:
-            return channels[component_id].new_receiver()
-
-        mocker.patch(
-            "frequenz.sdk.timeseries._formula_engine._resampled_formula_builder"
-            ".ResampledFormulaBuilder._get_resampled_receiver",
-            mock_resampled_receiver,
-        )
+        await mockgrid.start_mock_datapipeline(mocker)
 
         logical_meter = microgrid.logical_meter()
         grid_recv = logical_meter.grid_power.new_receiver()
         grid_production_recv = logical_meter.grid_production_power.new_receiver()
         grid_consumption_recv = logical_meter.grid_consumption_power.new_receiver()
 
-        now = datetime.now()
-        await send_resampled_data(now, (1.0, 2.0, 3.0, 4.0))
+        await mockgrid.mock_data.send_meter_power([1.0, 2.0, 3.0, 4.0])
         assert (await grid_recv.receive()).value == 10.0
         assert (await grid_production_recv.receive()).value == 0.0
         assert (await grid_consumption_recv.receive()).value == 10.0
 
-        await send_resampled_data(now, (1.0, 2.0, -3.0, -4.0))
+        await mockgrid.mock_data.send_meter_power([1.0, 2.0, -3.0, -4.0])
         assert (await grid_recv.receive()).value == -4.0
         assert (await grid_production_recv.receive()).value == 4.0
         assert (await grid_consumption_recv.receive()).value == 0.0
@@ -158,35 +121,7 @@ class TestLogicalMeter:
         mockgrid = MockMicrogrid(grid_side_meter=False)
         mockgrid.add_chps(1)
         mockgrid.add_batteries(2)
-        await mockgrid.start(mocker)
-
-        assert len(mockgrid.meter_ids) == 4
-
-        channels: dict[int, Broadcast[Sample]] = {
-            meter_id: Broadcast(f"#{meter_id}") for meter_id in [*mockgrid.meter_ids]
-        }
-        senders: list[Sender[Sample]] = [
-            channels[component_id].new_sender() for component_id in mockgrid.meter_ids
-        ]
-
-        async def send_resampled_data(
-            now: datetime,
-            meter_data: list[float | None],
-        ) -> None:
-            """Send resampled data to the channels."""
-            for sender, value in zip(senders, meter_data):
-                await sender.send(Sample(now, value))
-
-        def mock_resampled_receiver(
-            _1: Any, component_id: int, _2: ComponentMetricId
-        ) -> Receiver[Sample]:
-            return channels[component_id].new_receiver()
-
-        mocker.patch(
-            "frequenz.sdk.timeseries._formula_engine._resampled_formula_builder"
-            ".ResampledFormulaBuilder._get_resampled_receiver",
-            mock_resampled_receiver,
-        )
+        await mockgrid.start_mock_datapipeline(mocker)
 
         logical_meter = microgrid.logical_meter()
         chp_power_receiver = logical_meter.chp_power.new_receiver()
@@ -197,13 +132,12 @@ class TestLogicalMeter:
             logical_meter.chp_consumption_power.new_receiver()
         )
 
-        now = datetime.now(tz=timezone.utc)
-        await send_resampled_data(now, [1.0, 2.0, 3.0, 4.0])
+        await mockgrid.mock_data.send_meter_power([1.0, 2.0, 3.0, 4.0])
         assert (await chp_power_receiver.receive()).value == 2.0
         assert (await chp_production_power_receiver.receive()).value == 0.0
         assert (await chp_consumption_power_receiver.receive()).value == 2.0
 
-        await send_resampled_data(now, [-4.0, -12.0, None, 10.2])
+        await mockgrid.mock_data.send_meter_power([-4.0, -12.0, None, 10.2])
         assert (await chp_power_receiver.receive()).value == -12.0
         assert (await chp_production_power_receiver.receive()).value == 12.0
         assert (await chp_consumption_power_receiver.receive()).value == 0.0
@@ -212,36 +146,7 @@ class TestLogicalMeter:
         """Test the pv power formula."""
         mockgrid = MockMicrogrid(grid_side_meter=False)
         mockgrid.add_solar_inverters(2)
-        await mockgrid.start(mocker)
-
-        assert len(mockgrid.pv_inverter_ids) == 2
-
-        channels: dict[int, Broadcast[Sample]] = {
-            inv_id: Broadcast(f"#{inv_id}") for inv_id in [*mockgrid.pv_inverter_ids]
-        }
-        senders: list[Sender[Sample]] = [
-            channels[component_id].new_sender()
-            for component_id in mockgrid.pv_inverter_ids
-        ]
-
-        async def send_resampled_data(
-            now: datetime,
-            meter_data: list[float | None],
-        ) -> None:
-            """Send resampled data to the channels."""
-            for sender, value in zip(senders, meter_data):
-                await sender.send(Sample(now, value))
-
-        def mock_resampled_receiver(
-            _1: Any, component_id: int, _2: ComponentMetricId
-        ) -> Receiver[Sample]:
-            return channels[component_id].new_receiver()
-
-        mocker.patch(
-            "frequenz.sdk.timeseries._formula_engine._resampled_formula_builder"
-            ".ResampledFormulaBuilder._get_resampled_receiver",
-            mock_resampled_receiver,
-        )
+        await mockgrid.start_mock_datapipeline(mocker)
 
         logical_meter = microgrid.logical_meter()
         pv_power_receiver = logical_meter.pv_power.new_receiver()
@@ -250,8 +155,7 @@ class TestLogicalMeter:
             logical_meter.pv_consumption_power.new_receiver()
         )
 
-        now = datetime.now(tz=timezone.utc)
-        await send_resampled_data(now, [-1.0, -2.0])
+        await mockgrid.mock_data.send_meter_power([10.0, -1.0, -2.0])
         assert (await pv_power_receiver.receive()).value == -3.0
         assert (await pv_production_power_receiver.receive()).value == 3.0
         assert (await pv_consumption_power_receiver.receive()).value == 0.0
@@ -261,41 +165,12 @@ class TestLogicalMeter:
         mockgrid = MockMicrogrid(grid_side_meter=True)
         mockgrid.add_batteries(2)
         mockgrid.add_solar_inverters(2)
-        await mockgrid.start(mocker)
-
-        assert len(mockgrid.meter_ids) == 5
-
-        channels: dict[int, Broadcast[Sample]] = {
-            meter_id: Broadcast(f"#{meter_id}") for meter_id in [*mockgrid.meter_ids]
-        }
-        senders: list[Sender[Sample]] = [
-            channels[component_id].new_sender() for component_id in mockgrid.meter_ids
-        ]
-
-        async def send_resampled_data(
-            now: datetime,
-            meter_data: list[float | None],
-        ) -> None:
-            """Send resampled data to the channels."""
-            for sender, value in zip(senders, meter_data):
-                await sender.send(Sample(now, value))
-
-        def mock_resampled_receiver(
-            _1: Any, component_id: int, _2: ComponentMetricId
-        ) -> Receiver[Sample]:
-            return channels[component_id].new_receiver()
-
-        mocker.patch(
-            "frequenz.sdk.timeseries._formula_engine._resampled_formula_builder"
-            ".ResampledFormulaBuilder._get_resampled_receiver",
-            mock_resampled_receiver,
-        )
+        await mockgrid.start_mock_datapipeline(mocker)
 
         logical_meter = microgrid.logical_meter()
         consumer_power_receiver = logical_meter.consumer_power.new_receiver()
 
-        now = datetime.now(tz=timezone.utc)
-        await send_resampled_data(now, [20.0, 2.0, 3.0, 4.0, 5.0])
+        await mockgrid.mock_data.send_meter_power([20.0, 2.0, 3.0, 4.0, 5.0])
         assert (await consumer_power_receiver.receive()).value == 6.0
 
     async def test_consumer_power_no_grid_meter(self, mocker: MockerFixture) -> None:
@@ -303,39 +178,10 @@ class TestLogicalMeter:
         mockgrid = MockMicrogrid(grid_side_meter=False)
         mockgrid.add_batteries(2)
         mockgrid.add_solar_inverters(2)
-        await mockgrid.start(mocker)
-
-        assert len(mockgrid.meter_ids) == 5
-
-        channels: dict[int, Broadcast[Sample]] = {
-            meter_id: Broadcast(f"#{meter_id}") for meter_id in [*mockgrid.meter_ids]
-        }
-        senders: list[Sender[Sample]] = [
-            channels[component_id].new_sender() for component_id in mockgrid.meter_ids
-        ]
-
-        async def send_resampled_data(
-            now: datetime,
-            meter_data: list[float | None],
-        ) -> None:
-            """Send resampled data to the channels."""
-            for sender, value in zip(senders, meter_data):
-                await sender.send(Sample(now, value))
-
-        def mock_resampled_receiver(
-            _1: Any, component_id: int, _2: ComponentMetricId
-        ) -> Receiver[Sample]:
-            return channels[component_id].new_receiver()
-
-        mocker.patch(
-            "frequenz.sdk.timeseries._formula_engine._resampled_formula_builder"
-            ".ResampledFormulaBuilder._get_resampled_receiver",
-            mock_resampled_receiver,
-        )
+        await mockgrid.start_mock_datapipeline(mocker)
 
         logical_meter = microgrid.logical_meter()
         consumer_power_receiver = logical_meter.consumer_power.new_receiver()
 
-        now = datetime.now(tz=timezone.utc)
-        await send_resampled_data(now, [20.0, 2.0, 3.0, 4.0, 5.0])
+        await mockgrid.mock_data.send_meter_power([20.0, 2.0, 3.0, 4.0, 5.0])
         assert (await consumer_power_receiver.receive()).value == 20.0

